h2o-core/src/main/java/water/rapids/BinaryMerge.java

```diff
@@ -357,145 +351,134 @@ private long bsearchLeft(long x, boolean returnLow, long upp)  {
-    // i.e. (lLow+lUpp)/2 but being robust to one day in the future someone
-    // somewhere overflowing long; e.g. 32 exabytes of 1-column ints
-    long lr = lLow + (lUpp - lLow) / 2;
-    while (rLow < rUpp - 1) {
-      mid = rLow + (rUpp - rLow) / 2;
-      int cmp = keycmp(_leftKO._key, lr, _riteKO._key, mid);  // -1, 0 or 1, like strcmp
-      if (cmp < 0) {
-        rUpp = mid;
-      } else if (cmp > 0) {
-        rLow = mid;
-      } else { // rKey == lKey including NA == NA
-        // branch mid to find start and end of this group in this column
-        // TODO?: not if mult=first|last and col<ncol-1
-        tmpLow = mid;
-        tmpUpp = mid;
-        while (tmpLow < rUpp - 1) {
-          mid = tmpLow + (rUpp - tmpLow) / 2;
-          if (keycmp(_leftKO._key, lr, _riteKO._key, mid) == 0) tmpLow = mid;
-          else rUpp = mid;
+  /***
+   * For a specific MSB, if we can find MSBs in the left frame and rite frame that contains values in the MSB range,
+   * we will try to find match between the left frame and rite frame and include those rows in the final merged frame.
+   * If no match is found between the two frames but allLeft = true, we will go ahead and just include all the rows
+   * having the same MSB in the left frame into the final merged frame.
+   * 
+   * If allLeft is true, we will always iterate over the leftFrame keys and use binary search over the rite frame keys.
+   * We will count all duplicate keys in the left and rite frames as well.  During the search process, we should be 
+   * able to shrink the range of rite frame keys to search over as we gain more knowledge over the relative key 
+   * sizes between the two frames and all keys are sorted already.
+   * 
+   * However, when allLeft is false, we will iterate over the frames that has the smallest number of keys instead. 
+   * Hence, in this case, we can search over either the left or rite frame depending on the number of rows to be 
+   * searched through.  
+   * 
+   * @param leftLowIn: left frame lowest row number minus 1
+   * @param leftUppIn: number of rows in left frame with specific MSB
+   * @param riteLowIn: rite frame lowest row number minus 1 with specific MSB
+   * @param riteUppIn: number of rows in rite frame with specific MSB
+   *                  
+   */
+  private void binaryMergeOneMSB(long leftLowIn, long leftUppIn, long riteLowIn, long riteUppIn) {
+    if (!_allLeft && riteUppIn==0) return;  // no merging possible with empty rite frame here
+    boolean leftFrameIterate = _allLeft?true:((leftUppIn-leftLowIn)>(riteUppIn-riteLowIn)?false:true);
+    long iterIndex = leftFrameIterate?leftLowIn+1:riteLowIn+1;
+    long iterUppIn = leftFrameIterate?leftUppIn:riteUppIn;
+    long binarySearchUppIn = leftFrameIterate?riteUppIn:leftUppIn;
+    long iterHigh, binarySearchLow, binarySearchHigh, tempValue;  // temp variables to perform search
+    long[] midSearchInd = new long[3]; // store match row, riteLow, riteUpp
```

Comment:
there is no need for this to be an array, the comment is suggestion enough that this should be 3 variables or an object with 3 fields

Reply:
Honza: 

I made an object with three values out of it.  It makes things clearer and I still can get multiple values returned from functions.